Feature #week3

eeejung opened the pull request:
### **기능 세부 내용 구현 여부**


▶️ 관리자 페이지
- [x]  GET /adm/home/main
- [x]  관리자만 접근 가능
    ⇒ 회원 1명을 임의로 관리자로 지정 (`authLevel` == 7) → 해당 회원만 접근 가능
    ( user3 을 생성한 뒤 관리자로 지정하였습니다. )
    
<br>

▶️ 정산 데이터 생성
- [x]  정산 데이터 생성 폼 생성 및 처리 (`GET`, `POST`) → 관리자가 월별로 선택하여 생성
- [x]  결제 완료된 `orderItem` → 정산 아이템으로 생성

<br>

▶️ 정산
- [x]  정산 데이터 리스트 페이지
- [x]  건별 정산처리
    ⇒ 5대 5로 나눔, 수수료는 0원
- [x]  전체 정산처리 (여러개의 건들을 선택 → 전체 처리)

<br>

▶️ 배치 사용 (`추가 과제`)
- [ ]  정산데이터를 배치로 생성하기
    - 스프링 내장 스케쥴러를 이용해서 배치가 매달 15일 새벽 4시에 실행되도록
    - `@EnableScheduling` 사용
    - Quartz 사용금지, 스프링의 기본적인 스케쥴링 기능을 이용

<br>

▶️ 출금 기능 (`추가 과제`)
- [ ]  사용자 예치금 출금 신청 (금액, 통장, 계좌번호 입력)
- [ ]  관리자 페이지에서 출금 신청 목록 확인 가능
- [ ]  관리자가 출금 처리 (건별로 처리, 수수료는 0원)

---

- 저번 주말부터 이번 3주차 미션에는 개인 사정으로 시간을 많이 할애하지 못하여 완성도가 다소 떨어집니다.
- 스프링부트 음원결제 수업 코드를 최대한 활용하였습니다.
- ✏️ 3주차 Mission Record 링크 : [미션 레코드](https://github.com/likelion-backendschool/FinalProject_KimEunJung_team6/blob/master/3Week_Record/3Week_%EA%B9%80%EC%9D%80%EC%A0%95.md)

eeejung committed:
Feat : authLevel로 관리자 권한 설정, user3을 관리자로 지정

eeejung committed:
Comment : authLevel 주석 추가

eeejung committed:
Design : 관리자 홈 html 추가, layout 조정

eeejung committed:
Feat : 정산 데이터 생성 페이지 추가

eeejung committed:
Feat : 선택한 월에 맞는 정산 날짜 범위 생성

eeejung committed:
Feat : 정산 서비스 구현

eeejung committed:
Feat : RebateOrderItem에 구매자 정보, 정산금액지급내역번호 추가

eeejung committed:
Feat : rebateOrderItemList 구현

eeejung committed:
Refactor : 현재 날짜에 맞게 11월 정산 데이터 생성 추가

eeejung committed:
Feat : 건별 정산 처리 구현

eeejung committed:
Feat : 전체 정찬 처리 구현
체크박스 선택으로 전체 및 선택항목 정산 가능, 날짜 포맷 수정, 테스트를 위해 order initData 여러개 생성

eeejung committed:
Refactor : 5 대 5로 정산 계산 처리 수정

ghdtmdvy2 commented on 3Week_Mission/src/main/java/com/ll/exam/final__2022_10_08/app/rebate/controller/AdmRebateController.java:
```diff
@@ -0,0 +1,96 @@
+
+        RsData makeDateRsData = rebateService.makeDate(yearMonth);
+
+//        return Rq.redirectWithMsg("/adm/rebate/rebateOrderItemList", "%s월 정산데이터가 성공적으로 생성되었습니다.".formatted(yearMonth));
```
이 부분의 주석은 안지운거 같습니다!

eeejung replied:
헉 깜빡했네요 ㅠㅠ 감사합니다 !

ghdtmdvy2 commented on 3Week_Mission/src/main/java/com/ll/exam/final__2022_10_08/app/security/service/CustomUserDetailsService.java:
```diff
@@ -18,6 +22,10 @@ public class CustomUserDetailsService implements UserDetailsService {
     public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
         Member member = memberRepository.findByUsername(username).get();
 
+        if (member.getUsername().equals("user3")) {
+            member.setAuthLevel(7);     // user3을 관리자로 설정
+        }
+
```
사소한 거지만 이 부분은 `member.genAuthorities()` 안에서 처리해도 괜찮지 않을까요?

eeejung replied:
아하..! 승표님 말씀대로 그렇게 하는 방법이 더 깔끔할 것 같네요,, 추후 고쳐보겠습니다!

ghdtmdvy2 reviewed (commented):
되게 깔끔하게 코드가 정리되어있어서 보기 좋았습니다! 이번주도 수고하셨습니다!

shun-lee commented on 3Week_Mission/src/main/java/com/ll/exam/final__2022_10_08/app/member/entity/Member.java:
```diff
@@ -49,13 +51,19 @@ public String getJdenticon() {
         authorities.add(new SimpleGrantedAuthority("MEMBER"));
 
         // 닉네임을 가지고 있다면 작가의 권한을 가진다.
         if (StringUtils.hasText(nickname)) {
```
관리자 권한이 추가되어서 작가의 AuthLevel도 분리하는게 더 좋을 것 같아요!

eeejung replied:
엇 안그래도 승훈님 코드리뷰를 하면서 일반회원 작가회원의 AuthLevel도 따로 설정해주는게 좋을 것 같다고 생각했는데 ! !
리팩토링 할 때 수정해야겠네요 감사합니닷!

sh970901 commented on 3Week_Mission/src/main/java/com/ll/exam/final__2022_10_08/app/security/service/CustomUserDetailsService.java:
```diff
@@ -18,6 +22,10 @@ public class CustomUserDetailsService implements UserDetailsService {
     public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
         Member member = memberRepository.findByUsername(username).get();
 
+        if (member.getUsername().equals("user3")) {
```
user3을 생성할 때 바로 처리해도 좋을 것 같아요!

shun-lee reviewed (commented):
이번 주도 고생하셨습니다.!! 
코드가 보기 편리하여 좋았습니다. 잘 보고 갑니다 ^_^

jihun-24k commented on 3Week_Mission/src/main/java/com/ll/exam/final__2022_10_08/app/member/entity/Member.java:
```diff
@@ -49,13 +51,19 @@ public String getJdenticon() {
             authorities.add(new SimpleGrantedAuthority("AUTHOR"));
         }
 
+        // authLevel의 값이 7이라면 관리자 권한을 가진다.
+        if (authLevel == 7) {
+            authorities.add(new SimpleGrantedAuthority("ADMIN"));
+        }
+
```
저는 임의로 username 가지고 관리자를 설정해줬었는데 은정님 방식이 요구사항에 맞는거 같네요! 배워갑니다.

jihun-24k reviewed (commented):
코드 잘보고 갑니다😊
항상 Record를 잘 써주셔서 PR을 보다 쉽게 읽을 수 있었네요.
3주차도 고생많으셨습니다!!

hongjunpyo11 commented:
코드 잘 보고갑니다!! 정산 부분을 깔끔하게 구현을 정말 잘 해주신 것 같습니다!! 레코드에 이미지를 보여주며 리팩토링 방향을 보여주는 꼼꼼함도 배워갑니다!! 3주차도 고생하셨습니다!!

nette74 commented on 3Week_Mission/src/main/java/com/ll/exam/final__2022_10_08/app/member/entity/Member.java:
```diff
@@ -49,13 +51,19 @@ public String getJdenticon() {
         authorities.add(new SimpleGrantedAuthority("MEMBER"));
 
         // 닉네임을 가지고 있다면 작가의 권한을 가진다.
         if (StringUtils.hasText(nickname)) {
```
저는 일반유저, 작가, 관리자의 권한을 직선적으로 줬는데 작가의 권한은 별개로 주는게 낫겠군요..!

nette74 replied:
저는 일반유저, 작가, 관리자의 권한을 직선적으로 줬는데 작가의 권한은 별개로 주는게 낫겠군요..!

nette74 reviewed (commented):
저처럼 이번주에 시간을 많이 할애 하지 못 하셨는데도 기능 단단하게 잘 만드신 것 같습니다! 잘 보고 갑니다.

eeejung committed:
Refactor : 3주차 리팩터링
출금 기능 도입, 스프링 배치 도입 시도 (미완성), 일부 로직 수정, MyBook - 메서드 수정, Order - 주문 취소 + 환불 기능 도입
